presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometrySerde.java

```diff
@@ -358,58 +245,110 @@ private static void writePolyline(Geometry geometry, SliceOutput output, boolean
-            Coordinate[] coordinates = readCoordinates(input, partLengths[i]);
-            if (!isClockwise(coordinates)) {
-                holes.add(GEOMETRY_FACTORY.createLinearRing(coordinates));
-                continue;
-            }
-            if (shell != null) {
-                polygons.add(GEOMETRY_FACTORY.createPolygon(shell, holes.toArray(new LinearRing[0])));
-            }
-            shell = GEOMETRY_FACTORY.createLinearRing(coordinates);
-            holes.clear();
+    private static void writePolyline(Geometry geometry, SliceOutput output)
+    {
+        int numParts;
+        int numPoints = geometry.getNumPoints();
+        switch (geometry.getGeometryType()) {
+            case "MultiLineString":
+                numParts = geometry.getNumGeometries();
+                break;
+            case "LineString":
+                numParts = numPoints > 0 ? 1 : 0;
+                break;
+            default:
+                throw new IllegalArgumentException("Expected LineString or MultiLineString, but got " + geometry.getGeometryType());
         }
-        polygons.add(GEOMETRY_FACTORY.createPolygon(shell, holes.toArray(new LinearRing[0])));
 
-        if (polygons.size() == 1) {
-            return polygons.get(0);
+        switch (geometry.getGeometryType()) {
+            case "LineString":
+                output.writeByte(GeometryType.LINE_STRING.code());
+                break;
+            case "MultiLineString":
+                output.writeByte(GeometryType.MULTI_LINE_STRING.code());
+                break;
+            default:
+                throw new IllegalArgumentException("Expected LineString or MultiLineString, but got " + geometry.getGeometryType());
+        }
+
+        output.writeInt(EsriShapeType.POLYLINE.code);
+
+        writeEnvelope(geometry.getEnvelopeInternal(), output, geometry.isEmpty());
+
+        output.writeInt(numParts);
+        output.writeInt(numPoints);
+
+        int partIndex = 0;
+        for (int i = 0; i < numParts; i++) {
+            output.writeInt(partIndex);
+            partIndex += geometry.getGeometryN(i).getNumPoints();
         }
-        return GEOMETRY_FACTORY.createMultiPolygon(polygons.toArray(new Polygon[0]));
+
+        writeCoordinates(geometry.getCoordinates(), output);
     }
 
-    private static void writePolygon(Geometry geometry, SliceOutput output, boolean inGeometryCollection)
+    private static void writePolygon(Geometry geometry, SliceOutput output)
     {
-        requireNonNull(geometry, "geometry is null");
-        requireNonNull(output, "output is null");
-        if (!geometry.getGeometryType().equals(JTS_POLYGON) && !geometry.getGeometryType().equals(JTS_MULTI_POLYGON)) {
+        if (!geometry.getGeometryType().equals("Polygon") && !geometry.getGeometryType().equals("MultiPolygon")) {
             throw new IllegalArgumentException("Expected LineString or MultiLineString, but got " + geometry.getGeometryType());
```

Comment:
s/LineString/Polygon

s/MultiLineString/MultiPolygon

Reply:
I just removed the check